src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/DownstreamBuildAction.java

```diff
@@ -6,47 +6,80 @@
 import hudson.model.Item;
 import hudson.model.ItemGroup;
 import hudson.model.Run;
-import org.jenkinsci.plugins.workflow.actions.PersistentAction;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
 import org.springframework.security.access.AccessDeniedException;
 
 /**
- * Tracks the downstream build triggered by the {@code build} step.
+ * Tracks downstream builds triggered by the {@code build} step, as well as the {@link FlowNode#getId} of the step.
  *
  * @see BuildUpstreamCause
  */
-public final class DownstreamBuildAction extends InvisibleAction implements PersistentAction {
-    private final String jobFullName;
-    private Integer buildNumber;
+public final class DownstreamBuildAction extends InvisibleAction {
+    private final Map<String, DownstreamBuild> downstreamBuilds = new LinkedHashMap<>();
 
-    DownstreamBuildAction(Item job) {
-        this.jobFullName = job.getFullName();
+    public static @NonNull DownstreamBuild getOrCreate(@NonNull Run<?, ?> run, @NonNull String flowNodeId, @NonNull Item job) {
+        DownstreamBuildAction downstreamBuildAction;
+        synchronized (DownstreamBuildAction.class) {
```

Comment:
Unnecessary in the normal case since the relevant part of `BuildTriggerStepExecution` runs synchronously on the CPS VM thread, but if there are queued triggered builds when updating Jenkins to pick up this PR or when Jenkins crashes then there is a possibility of race conditions leading to multiple instances of `DownstreamBuildAction` without some kind of synchronization.

Reply:
May as well keep it, does not add much clutter.